SWIPs/swip-forwarder.md

```diff
@@ -0,0 +1,181 @@
+
+
+## Abstract
+
+Currently all components that need to send messages to peers make their own calculations to determine which peer to send to. In particular the retrieve, push sync and pss protocols all need this feature.
+
+This SWIP proposes to provide this feature as a standard service of the network package.
+
+## Motivation
+
+Avoid parallell implementations of the same feature.
+
+## Specification
+
+The reference implementation is specified in `golang`.
+
+```
+
+New(capabilityIndex) SessionInterface
+
+
+type SessionInterface interface {
+	Subscribe() <-ForwardPeer
+	Get(numberOfPeers int) ([]ForwardPeer, error)
+	Close()
+}
+
+// also implements context.Context
+type SessionContext struct {
+	CapabilityIndex string
+	SessionId string
+}
+
+type SessionRPCInterface interface {
+	NewForwardPeerSession(ctx context.Context) (context.Context, error)
+	Subscribe(ctx context.Context, "kademlia", ch chan PeerEvent, "forward") (sub rpc.Subscription, err error)
+	GetForwardPeer(ctx context.Context, numberOfPeers int) (peers []ForwardPeer, err error)
+	CloseForwardPeerSession(ctx context.Context)
+}
+```
+
+### Library interface
+
+The `capabilityIndex` argument to `New` and `CapabilityIndex` member of `SessionContext` refers to the key of a  `Capability` object registered with `Kademlia.RegisterCapabilityIndex`. Provide an empty string means peers will not be filtered by capability.
+
+`Get` provides two modes of operation:
+
+* If `Subscribe` has not been called, `Get(n)` returns a slice of up to `n` peers.
+* If `Subscribe` has been called, `Get(n)` returns an empty slice, and up to `n` peers will be put on the channel instead.
+
+`Get` called with `numberOfPeers = 0` returns _all_ matching peers in order.
+
+`Close` _MUST_ always be called to free up the component's session peer cache/cursor, and the subscription channel if subscription has been made.
+
+### RPC interface
+
+If `Get` is called with `SessionContext` _and_ the `SessionId` variable is set _and_ matches a `Session` created through `rpc.Subscribe`, then the next eligible peers for the `Session` will be returned on the subscription channel.
+
+Otherwise, the `Session` terminates immediately after completing the `Get` request.
+
+To initiate a `Session`, `NewForwardPeerSession` should be called with a `SessionContext` as the parameter (a different context will generate an error). The returned context will contain a unique `sessionId` key generated by the server. This context can in turn be passed to `Get` and (optionally) `Subscribe` to attain the corresponding functionality provided by the library interface.
+
+If `NewForwardPeerSession` has been called, `CloseForwardPeerSession` _MUST_ be called with the `SessionContext` updated by the server to terminate the `Session`.
+
+## Rationale
+
+The component shall provide a service where given a specific Swarm Overlay Address, connected peers will be returned in the order of closest to farthest.
```

Comment:
you mention a `given Swarm Overlay Address`, however a pivot address is not mentioned in any of the function signatures you've defined. In layman terms: you need to know which peer to forward to given a content addressed hash and the peers to forward to would be ordered according to that reference, but that reference parameter is not mentioned in the signatures (or am i getting this all wrong?)